### ISSUE TITLE
xterm: last char in first line changes to "2" on file open

### ISSUE THREAD

**Issue by rysson** [drmikehenry/vim-fixkey#11]
Hi,
it's awesome plugin!

I have one annoying bug. When I open a file (`vim` or `vim -R`) always last character in first line is changed to "`2`". When first line is empty nothing changes.
*EDIT:* Next (in both cases) cursor is moved to second line.
**EDIT2:** Affect only on `xterm`.

I tested on empty installation. Only Debian default ViM installed plus VundleVim and vim-fixkey.

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jun 15 2019 16:41:15)
Included patches: 1-875, 878, 884, 948, 1046, 1365-1368, 1382, 1401
Modified by team+vim@tracker.debian.org
Compiled by team+vim@tracker.debian.org
Huge version with GTK2 GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     +tag_old_static
+arabic            +farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    +tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
+balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
+browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +textprop
+byte_offset       -hangul_input      +num64             +timers
+channel           +iconv             +packages          +title
+cindent           +insert_expand     +path_extra        +toolbar
+clientserver      +job               +perl              +user_commands
+clipboard         +jumplist          +persistent_undo   +vartabs
+cmdline_compl     +keymap            +postscript        +vertsplit
+cmdline_hist      +lambda            +printer           +virtualedit
+cmdline_info      +langmap           +profile           +visual
+comments          +libcall           -python            +visualextra
+conceal           +linebreak         +python3           +viminfo
+cryptv            +lispindent        +quickfix          +vreplace
+cscope            +listcmds          +reltime           +wildignore
+cursorbind        +localmap          +rightleft         +wildmenu
+cursorshape       +lua               +ruby              +windows
+dialog_con_gui    +menu              +scrollbind        +writebackup
+diff              +mksession         +signs             +X11
+digraphs          +modify_fname      +smartindent       -xfontset
+dnd               +mouse             +startuptime       +xim
-ebcdic            +mouseshape        +statusline        +xpm
+emacs_tags        +mouse_dec         -sun_workshop      +xsmp_interact
+eval              +mouse_gpm         +syntax            +xterm_clipboard
+ex_extra          -mouse_jsbterm     +tag_binary        -xterm_save
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-2.0 -I/usr/lib/x86_64-linux-gnu/gtk-2.0/include -I/usr/include/gio-unix-2.0 -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -Wdate-time  -g -O2 -fdebug-prefix-map=/build/vim-4Pursk/vim-8.1.0875=. -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: gcc   -L. -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-x11-2.0 -lgdk-x11-2.0 -lpangocairo-1.0 -latk-1.0 -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lpangoft2-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lfontconfig -lfreetype -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.28/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.7/config-3.7m-x86_64-linux-gnu -lpython3.7m -lcrypt -lpthread -ldl -lutil -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lm -lruby-2.5 -lpthread -lgmp -ldl -lcrypt -lm
```

**Comment by drmikehenry** [comment]
Which version of Debian are you using?  I think you're the first person to mention this problem.  It might be something in the Debian-supplied vimrc settings.  Which terminal emulator are you using, and what does `:echo $TERM` show in Vim?

**Comment by rysson** [comment]
Thanks for so quickly answer. Yes, I'm surprised too :)

Debian Buster (stable). The same is in Debian Bullseye (testing).
I'm using `xterm` (via `uxterm` script). `$TERM` is `xterm-256color`

`/usr/share/vim/vim81/debian.vim`:
```vim
" Debian system-wide default configuration Vim

set runtimepath=~/.vim,/var/lib/vim/addons,/usr/share/vim/vimfiles,/usr/share/vim/vim81,/usr/share/vim/vimfiles/after,/var/lib/vim/addons/after,~/.vim/after

" Normally we use vim-extensions. If you want true vi-compatibility
" remove change the following statements
set nocompatible	" Use Vim defaults instead of 100% vi compatibility
set backspace=indent,eol,start	" more powerful backspacing

" Now we set some defaults for the editor
set history=50		" keep 50 lines of command line history
set ruler		" show the cursor position all the time

" modelines have historically been a source of security/resource
" vulnerabilities -- disable by default, even when 'nocompatible' is set
set nomodeline

" Suffixes that get lower priority when doing tab completion for filenames.
" These are files we are not likely to want to edit or read.
set suffixes=.bak,~,.swp,.o,.info,.aux,.log,.dvi,.bbl,.blg,.brf,.cb,.ind,.idx,.ilg,.inx,.out,.toc

" We know xterm-debian is a color terminal
if &term =~ "xterm-debian" || &term =~ "xterm-xfree86"
  set t_Co=16
  set t_Sf=^[[3%dm
  set t_Sb=^[[4%dm
endif

" Some Debian-specific things
if has('gui')
  " Must define this within the :if so it does not cause problems with
  " vim-tiny (which does not have +eval)
  function! <SID>MapExists(name, modes)
    for mode in split(a:modes, '\zs')
      if !empty(maparg(a:name, mode))
        return 1
      endif
    endfor
    return 0
  endfunction

  " Make shift-insert work like in Xterm
  autocmd GUIEnter * if !<SID>MapExists("<S-Insert>", "nvso") | execute "map <S-Insert> <MiddleMouse>" | endif
  autocmd GUIEnter * if !<SID>MapExists("<S-Insert>", "ic") | execute "map! <S-Insert> <MiddleMouse>" | endif
endif

" Set paper size from /etc/papersize if available (Debian-specific)
if filereadable("/etc/papersize")
  let s:papersize = matchstr(readfile('/etc/papersize', '', 1), '\p*')
  if strlen(s:papersize)
    exe "set printoptions+=paper:" . s:papersize
  endif
endif
```

There are also some plugins (from `/usr/share/vim/vim81/plugin/README.txt`):
```
The plugin directory is for standard Vim plugin scripts.

All files here ending in .vim will be sourced by Vim when it starts up.
Look in the file for hints on how it can be disabled without deleting it.

getscriptPlugin.vim  get latest version of Vim scripts
gzip.vim       edit compressed files
logiPat.vim      logical operators on patterns
manpager.vim       using Vim as manpager
matchparen.vim       highlight paren matching the one under the cursor
netrwPlugin.vim      edit files over a network and browse (remote) directories
rrhelper.vim       used for --remote-wait editing
spellfile.vim      download a spellfile when it's missing
tarPlugin.vim      edit (compressed) tar files
tohtml.vim       convert a file with syntax highlighting to HTML
vimballPlugin.vim    create and unpack .vba files
zipPlugin.vim      edit zip archives

Note: the explorer.vim plugin is no longer here, the netrw.vim plugin has
taken over browsing directories (also for remote directories).
```

Plugins can be download from Debian site: https://packages.debian.org/bullseye/vim-runtime.

**Comment by drmikehenry** [comment]
I'm on my way to work, but here's a couple of things you can try.  Use `:scriptnames` to see which files have been sourced.  You may want to rename the `/usr/share/vim/vim81/plugin` directory out of the way to prevent those scripts from loading, and similarly for `/etc/vim/vimrc` or any other scripts.  There are also switches like `vim -u NONE` and `vim --clean`, but that will prevent `fixkey` from loading in the normal way.  It also would be worth checking with another terminal emulator or two.

**Comment by rysson** [comment]
Sure. It's what I found.

## Scripts

Result of `:scriptnames`
```
  1: /usr/share/vim/vimrc
  2: /usr/share/vim/vim81/debian.vim
  3: ~/.vimrc
  4: /usr/share/vim/vim81/syntax/syntax.vim
  5: /usr/share/vim/vim81/syntax/synload.vim
  6: /usr/share/vim/vim81/syntax/syncolor.vim
  7: /usr/share/vim/vim81/filetype.vim
  8: /usr/share/vim/vim81/ftoff.vim
  9: ~/.vim/bundle/Vundle.vim/autoload/vundle.vim
 10: ~/.vim/bundle/Vundle.vim/autoload/vundle/config.vim
 11: /usr/share/vim/vim81/ftplugin.vim
 12: /usr/share/vim/vim81/indent.vim
 13: ~/.vim/bundle/vim-fixkey/plugin/fixkey.vim
 14: /usr/share/vim/vim81/plugin/getscriptPlugin.vim
 15: /usr/share/vim/vim81/plugin/gzip.vim
 16: /usr/share/vim/vim81/plugin/logiPat.vim
 17: /usr/share/vim/vim81/plugin/manpager.vim
 18: /usr/share/vim/vim81/plugin/matchparen.vim
 19: /usr/share/vim/vim81/plugin/netrwPlugin.vim
 20: /usr/share/vim/vim81/plugin/rrhelper.vim
 21: /usr/share/vim/vim81/plugin/spellfile.vim
 22: /usr/share/vim/vim81/plugin/tarPlugin.vim
 23: /usr/share/vim/vim81/plugin/tohtml.vim
 24: /usr/share/vim/vim81/plugin/vimballPlugin.vim
 25: /usr/share/vim/vim81/plugin/zipPlugin.vim
 26: /usr/share/vim/vim81/scripts.vim
```

Testing `.vimrc`:
```vim
syn on

set nocompatible              " be iMproved, required
filetype off                  " required

" set the runtime path to include Vundle and initialize
set rtp+=~/.vim/bundle/Vundle.vim
call vundle#begin()

Plugin 'VundleVim/Vundle.vim'
Plugin 'drmikehenry/vim-fixkey'

call vundle#end()            " required
filetype plugin indent on    " required
```

Scripts without system plugins.
```
  1: /usr/share/vim/vimrc
  2: /usr/share/vim/vim81/debian.vim
  3: ~/.vimrc
  4: /usr/share/vim/vim81/syntax/syntax.vim
  5: /usr/share/vim/vim81/syntax/synload.vim
  6: /usr/share/vim/vim81/syntax/syncolor.vim
  7: /usr/share/vim/vim81/filetype.vim
  8: /usr/share/vim/vim81/ftoff.vim
  9: ~/.vim/bundle/Vundle.vim/autoload/vundle.vim
 10: ~/.vim/bundle/Vundle.vim/autoload/vundle/config.vim
 11: /usr/share/vim/vim81/ftplugin.vim
 12: /usr/share/vim/vim81/indent.vim
 13: ~/.vim/bundle/vim-fixkey/plugin/fixkey.vim
 14: /usr/share/vim/vim81/scripts.vim
```

After remove almost all system scripts.
```
  1: /usr/share/vim/vimrc
  2: ~/.vimrc
  3: ~/.vim/bundle/Vundle.vim/autoload/vundle.vim
  4: ~/.vim/bundle/Vundle.vim/autoload/vundle/config.vim
  5: ~/.vim/bundle/vim-fixkey/plugin/fixkey.vim
```

Still the same effect.

## Terminals

| Terminal emulator | OK |
| --- | --- |
| console (linux VT) | + |
| deepin-terminal | + |
| Eterm | + |
| gnome-terminal | + |
| kitty | + |
| konsole | + |
| kterm | + |
| lxterminal | + |
| mate-terminal | + |
| mrxvt | + |
| rxvt | + |
| stterm | + |
| termit | + |
| tilda | + |
| tilix | + |
| xterm | – |

Damn it, what is wrong with my favorite terminal?

I've removed all x-resources, the same effect.

**Comment by drmikehenry** [comment]
I wonder if there is something going on with the terminal response string.  I seem to recall some funny characters showing up at Vim startup for various terminals.  Here are a couple of links to consider:
- https://vi.stackexchange.com/questions/14491/strange-vim-xterm-interaction-auto-insertion-of-register-when-opening-file
- https://github.com/vim/vim/issues/2032
- http://vim.1045645.n5.nabble.com/stray-control-characters-appear-on-vim-startup-td5730406.html

I'm not sure if any of the advice in the above links will be helpful in narrowing down what's going on.  You've done a good job of isolating it down to just `/usr/share/vim/vimrc`, your `.vimrc`, and the plugin manager.  You could get rid of Vundle as well and unpack `fixkey` directly into `~/.vim` to reduce the problem to just `fixkey`, `xterm` with its settings, and the Debian-supplied build of Vim.  (In other words, instead of `~/.vim/bundle/vim-fixkey/plugin/fixkey.vim`, you'd eliminate the `bundle/vim-fixkey/` directories and have `~/.vim/plugin/fixkey.vim`, and Vim would know to source that file at startup.)  You might also try creating a new user account on your system and running from there.  That would eliminate any per-user configuration settings that might be causing problems.

**Comment by rysson** [comment]
Wow. Very interesting but didn't help too much.
Yes, I'm working on new user. All registers have no `2`.

I've looked into plugin and comment (almost) random lines ;-)
I've found that problem occurs in `Fixkey_setMetaLetters()` for `P` key.
If I add `if c != 'p'` in loop every thing is working fine.

I don't know what I could break else, but this fix is working for me at that moment:
```diff
diff --git a/plugin/fixkey.vim b/plugin/fixkey.vim
index a8d37d2..dd24d3e 100644
--- a/plugin/fixkey.vim
+++ b/plugin/fixkey.vim
@@ -79,7 +79,8 @@ function! Fixkey_setMetaLetters()
         let uc = toupper(c)
         call Fixkey_setKey("<M-" .  c . ">", "\e" . c)
         " Since many keycodes have "\eO" in them, we can't use "\eO" for <M-O>.
-        if uc != 'O'
+        " <M-P> breaks something in xterm, see https://github.com/drmikehenry/vim-fixkey/issues/11.
+        if uc != 'O' && uc != 'P'
             call Fixkey_setKey("<M-" . uc . ">", "\e" . uc)
         endif
         let c = nr2char(char2nr(c) + 1)
```

I know, that fix *too much*, I should add terminal detection. At least
```vim
let skip_p = $TERM =~# '^xterm\(-\d*color\)\?$' && $COLORTERM != "gnome-terminal"
...
if uc != 'O' && (uc != 'P' || ! skip_p)
```

It's still too much, it affect gnome-terminal etc. I need read more about `v:termresponse`.
My xterm returns `^[[>41;348;0c`.

**Comment by rysson** [comment]
`set t_RV=` in `~/.vimrc` helps too. 
Maybe it's some kind of racing of control sequences?

**Comment by rysson** [comment]
I can't use `v:termresponse` in plugin because it's not set yet.

I give up. I set a global variable `g:Fixkey_xterm_fix` to force skip `<M-P>`, it have to be set in `~/.vimrc` before you load the plugin.

Current patch.
```diff
diff --git a/plugin/fixkey.vim b/plugin/fixkey.vim
index a8d37d2..e27eb7a 100644
--- a/plugin/fixkey.vim
+++ b/plugin/fixkey.vim
@@ -74,12 +74,14 @@ function! Fixkey_resetMetaNumbers()
 endfunction
 
 function! Fixkey_setMetaLetters()
+    let skip_p = exists("g:Fixkey_xterm_fix") && g:Fixkey_xterm_fix
     let c = 'a'
     while c <= 'z'
         let uc = toupper(c)
         call Fixkey_setKey("<M-" .  c . ">", "\e" . c)
         " Since many keycodes have "\eO" in them, we can't use "\eO" for <M-O>.
-        if uc != 'O'
+        " <M-P> breaks something in xterm, see https://github.com/drmikehenry/vim-fixkey/issues/11.
+        if uc != 'O' && (uc != 'P' || ! skip_p)
             call Fixkey_setKey("<M-" . uc . ">", "\e" . uc)
         endif
         let c = nr2char(char2nr(c) + 1)
```

If you could handle this issue in better way I will be happy :-)

**Comment by drmikehenry** [comment]
@rysson I think you've narrowed things down greatly.  I suspect Vim is still waiting for a response from Xterm while it's allowing Fixkey to initialize; then later, when the response arrives, Vim interprets Xterm's response through the lens of the new key definitions.  I think this happens only at startup, and only for Xterm because there is a fancier terminal interrogation that goes on for Xterm when the initial `v:termresponse` indicates an Xterm version of 141 or higher (see `:help xterm-codes` in Vim's help).  Most other terminals apparently return a version of 95 (the version where Xterm first grew the version interrogation feature; see `:help v:termresponse`).  I further suspect that delaying Fixkey initialization at startup might fix things.  I've pushed a branch named `delay` with some changes to delay by 100ms at startup (configurable via `g:Fixkey_setupDelay`; see the branch for detail).  If this fixes your problem, I'll make a new release.

**Comment by rysson** [comment]
Great, thanks a lot.

Report: It works! 
It's enough even if g:Fixkey_setupDelay is 1. No delay (zero) makes issue again.
I think it have to switch thread and give any chance to receive xterm response.
You can change initial value to 10 safely, IMO.

I don't use Neovim to much. I just test it time to time. 
I confirm, now there is no errors for Neovim.

I suggest a line of description about g:Fixkey_setupDelay in README.

Thank you very much for you work.

**Comment by drmikehenry** [comment]
I'm glad to hear it works for you.  I've released version 0.3.11 containing the startup delay feature and detection of Neovim.  Thanks for your help diagnosing this problem and testing the results!
